Introduce Fortran Test Framework and CTest Integration

amstokely opened the pull request:
**Summary:**
This PR introduces a Fortran testing framework and CTest integration to improve the testing and validation process for the library. Key updates include an assertion module, unit tests for core features, and CMake functions for streamlined test management. To test these changes, run ctest in the build directory after compiling obs2ioda.

st-ncar commented on obs2ioda-v2/README.md:
````diff
@@ -22,11 +22,11 @@ If you have an environment preconfigured for `mpas-jedi`, simply source that env
    find . -name *libbufr*
    ```
-4. Next, run CMake to configure the build. Remember to specify the path to the NCEP BUFR library:
+4. Next, run CMake to configure the build. Specify the `CMAKE_BUILD_TYPE` option to set the build type. Currently, the supported types are `Release`, `RelWithDebInfo`, and `Debug`. Don't forget to include the path to the NCEP BUFR library:
````
Thank you for the clear install instructions, they are great! The build types 'Release' and 'Debug' seem intuitive, but it is less clear to me under what circumstances I should use 'RelWithDebInfo' instead of say 'Release'. Would it be worthwhile to add a comment on the use cases for the build types?

amstokely replied:
I added a table that Summarizes the properties and use cases for each supported CMake build type in the latest commit

jim-p-w commented on obs2ioda-v2/src/test/fortran_test_framework_mod.f90:
```diff
@@ -0,0 +1,195 @@
+module fortran_test_framework_mod
+    implicit none
+
+    public :: assertEqual
+    public :: assert_interface
+    public :: determine_test_type
+    public :: assert
+    public :: assert_memcheck
```
public is the default access. You need to use `private` if you want parts of the module to be private.

jim-p-w commented on obs2ioda-v2/src/test/fortran_test_framework_mod.f90:
```diff
@@ -0,0 +1,195 @@
+module fortran_test_framework_mod
```
Having the word `fortran` as part of the module names (and file names) seems redundant and a bit cumbersome.

amstokely replied:
Agree. Do you think test_utils is a better name?

jim-p-w commented:
I don't understand the file hierarchy or the file nomenclature.
Can you add comments to the files explaining what they are?

Is `test/fortran/Test_fortran_test_framework_driver.f90` an actual test? if so I would omit `driver` from the file name.

Why is `test/fortran/Test_fortran_test_framework_mod.f90` in `test/fortran` and not in `src/test`?  It looks like it provides utility functions to be used by tests?

amstokely commented on obs2ioda-v2/src/test/fortran_test_framework_mod.f90:
```diff
@@ -0,0 +1,195 @@
+        end if
+    end subroutine assert
+
+    ! Subroutine: assert_memcheck
```
That's a great point. I change the name to `log_condition`

st-ncar commented on obs2ioda-v2/src/test/test_utils.f90:
```diff
@@ -0,0 +1,263 @@
+        assert_handler => assert_procedure
+
+        call assert_handler(&
+                expected == actual, "expected='" // trim(expected) // "' actual='" // &
```
What is the expected behavior if the string variables expected and actual have leading spaces? At current, ' a' and 'a' would not be considered equal, but 'a ' and 'a' would. Maybe it would be more consistent if leading and trailing spaces were both ignored in the comparison? In this case, we should left-adjust and trim the expected and actual string before the comparison in assert_handler.

st-ncar commented on obs2ioda-v2/src/test/test_utils.f90:
```diff
@@ -0,0 +1,263 @@
+        if (.not. condition) then
+            status = 1
+            write(*, '(A)') "Failed: " // message
+            stop 1
```
This is a stylistic comment and can also be ignored. I find the distinction between assert and assert_log, and their signature enforced by the assert_interface a bit confusing. It seems like assert_log provides the basic functionality and assert could in principle call assert_log and terminate program execution if status == 1 (that way its status wouldn't be unused). I think we talked a bit about this during our chat and I forgot our conclusion, sorry if this is a duplicate. And as I said, this is stylistic, I think assert_log and assert work fine as is.

amstokely replied:
I agree that assert should have a suffix given that assert_log does. I'll update this in my next commit

amstokely commented:
> I don't understand the file hierarchy or the file nomenclature.
> 
> Can you add comments to the files explaining what they are?
> 
> 
> 
> Is `test/fortran/Test_fortran_test_framework_driver.f90` an actual test? if so I would omit `driver` from the file name.
> 
> 
> 
> Why is `test/fortran/Test_fortran_test_framework_mod.f90` in `test/fortran` and not in `src/test`?  It looks like it provides utility functions to be used by tests?

I agree that the file naming/structure is confusing. I changed the name of the module from fortran_test_framework to test_utils and moved the test program and test module into one file, which should significantly clean up the overall design of the new module.
